docs/current_docs/cookbook/snippets/execute-concurrently/python/main.py

```diff
@@ -0,0 +1,60 @@
+import anyio
+
+import dagger
+from dagger import dag, function, object_type
+
+
+@object_type
+class MyModule:
+    @function
+    async def test(self, source: dagger.Directory) -> str:
+        """Return the result of running unit tests"""
+        return await (
+            self.build_env(source)
+            .with_exec(["npm", "run", "test:unit", "run"])
+            .stdout()
+        )
+
+    @function
+    async def typecheck(self, source: dagger.Directory) -> str:
+        """Return the result of running the type checker"""
+        return await (
+            self.build_env(source).with_exec(["npm", "run", "type-check"]).stdout()
+        )
+
+    @function
+    async def lint(self, source: dagger.Directory) -> str:
+        """Return the result of running the linter"""
+        return await (
+            self.build_env(source).with_exec(["npm", "run", "lint"]).stdout()
+        )
+
+    @function
+    async def run_all_tests(self, source: dagger.Directory) -> str:
+        """Run linter, type-checker, unit tests concurrently"""
+        # create task group
+        async with anyio.create_task_group() as tg:
+            lint_op = tg.start_soon(self.lint, source)
+            typecheck_op = tg.start_soon(self.typecheck, source)
+            test_op = tg.start_soon(self.test, source)
+
+        # wait for all tests to complete
+        lint_result = await lint_op
+        typecheck_result = await typecheck_op
+        test_result = await test_op
+
+        # if all tests succeed, print the test results
+        return lint_result + "\n" + typecheck_result + "\n" + test_result
+
```

Comment:
The [`TaskGroup.start_soon`](https://anyio.readthedocs.io/en/stable/api.html#anyio.abc.TaskGroup.start_soon) function returns `None`, not a `Future`, so `await lint_op` is resolving to `await None`. The context manager (`async with .... as tg`) already waits for those tasks to complete, so by the time the line with `await lint_op` executes, all tasks have already completed.

I would simplify this a little bit and not concatenate the results, just rely on TUI/Cloud UI:
```suggestion
    async def run_all_tests(self, source: dagger.Directory):
        """Run linter, type-checker, unit tests concurrently"""
        async with anyio.create_task_group() as tg:
            tg.start_soon(self.lint, source)
            tg.start_soon(self.typecheck, source)
            tg.start_soon(self.test, source)
```

Alternatively, it's a common pattern to have those functions return a `dagger.Container` rather than the `stdout` output because that gives you flexibility when calling them in the CLI for example. In that case, they wouldn't be async and this function could call them like this:
```python
async with anyio.create_task_group() as tg:
    tg.start_soon(self.lint(source).sync)
    tg.start_soon(self.typecheck(source).sync)
    tg.start_soon(self.test(source).sync)
```

In case of any errors, you can let the exception bublle up and it should be visible in the stack trace what those errors were (if you find that they aren't, please let me know). But to handle errors here, you just wrap the task group in a try...except:

```python
try:
    async with anyio.create_task_group() as tg:
        tg.start_soon(self.lint(source).sync)
        tg.start_soon(self.typecheck(source).sync)
        tg.start_soon(self.test(source).sync)
except* ValueError as excgroup:
    for exc in excgroup.exceptions:
        ...  # handle each ValueError
except* KeyError as excgroup:
    for exc in excgroup.exceptions:
        ...  # handle each KeyError
```
Of course that depends on which exceptions you want to catch or handle, but since in Go you have this:
```go
	if err := eg.Wait(); err != nil {
		return "", fmt.Errorf("error: %w", err)
	}
```
It's not doing any handling, so the equivalent in Python (and TypeScript) is to just let the exception bubble up the stack.

If in any case you really want to save those results (I suggest you don't), the `anyio` library doesn't do that in a simple way like you were trying to do. Some people use the [asyncer](https://asyncer.tiangolo.com/tutorial/soonify-return/#get-the-return-value-from-soonvalue-objects) library for that. But the cannonical way with anyio would be the use a [stream](https://anyio.readthedocs.io/en/stable/streams.html#text-streams), so rather than putting each result in their own variable and concatenate later, you'd send the results to a stream as they come, and then consume them in the end. Equivalent to a Go channel.

Reply:
I updated this as well @helderco as below:

```
import anyio

import dagger
from dagger import dag, function, object_type


@object_type
class MyModule:
    source: dagger.Directory

    @function
    async def test(self) -> str:
        """Return the result of running unit tests"""
        return await (
            self.build_env()
            .with_exec(["npm", "run", "test:unit", "run"])
            .stdout()
        )

    @function
    async def typecheck(self) -> str:
        """Return the result of running the type checker"""
        return await (
            self.build_env().with_exec(["npm", "run", "type-check"]).stdout()
        )

    @function
    async def lint(self) -> str:
        """Return the result of running the linter"""
        return await (
            self.build_env().with_exec(["npm", "run", "lint"]).stdout()
        )

    @function
    async def run_all_tests(self):
        """Run linter, type-checker, unit tests concurrently"""
        async with anyio.create_task_group() as tg:
            tg.start_soon(self.lint)
            tg.start_soon(self.typecheck)
            tg.start_soon(self.test)

    @function
    def build_env(self) -> dagger.Container:
        """Build a ready-to-use development environment"""
        node_cache = dag.cache_volume("node")
        return (
            dag.container()
            .from_("node:21-slim")
            .with_directory("/src", self.source)
            .with_mounted_cache("/root/.npm", node_cache)
            .with_workdir("/src")
            .with_exec(["npm", "install"])
        )
```

```
dagger call --source=. run-all-tests 
✔ connect 0.3s
✔ loading module 12.8s
✔ parsing command line arguments 0.2s
✔ myModule(
    source: ✔ ModuleSource.resolveDirectoryFromCaller(path: "."): Directory! 0.0s
  ): MyModule! 2.1s
✔ MyModule.runAllTests: Void 1m11.5s
```

There is an error when linting standalone though:

```
dagger call --source=. lint
✔ connect 0.8s
✔ loading module 2.1s
✔ parsing command line arguments 0.1s
✔ myModule(
    source: ✔ ModuleSource.resolveDirectoryFromCaller(path: "."): Directory! 0.0s
  ): MyModule! 3.2s
✔ MyModule.lint: String! 2.6s

Full trace at https://dagger.cloud/dagger/traces/77b74004c16948e3e954a7c909cf7bca


> hello-dagger@0.0.0 lint
> eslint . --ext .vue,.js,.jsx,.cjs,.mjs,.ts,.tsx,.cts,.mts --fix --ignore-path .gitignore


/src/src/App.vue
  2:10  warning  'RouterLink' is defined but never used  @typescript-eslint/no-unused-vars
  2:22  warning  'RouterView' is defined but never used  @typescript-eslint/no-unused-vars

✖ 2 problems (0 errors, 2 warnings)
```

Reply:
After this change, here is the output

```
dagger call --source=. run-all-tests
✔ connect 0.6s
✔ loading module 4.3s
✔ parsing command line arguments 0.3s
✔ myModule(
    source: ✔ ModuleSource.resolveDirectoryFromCaller(path: "."): Directory! 0.0s
  ): MyModule! 3.5s
✘ MyModule.runAllTests: Void 56.4s
! call function "run_all_tests": process "/runtime" did not complete successfully: exit code: 1
┃ ╭─ Error ──────────────────────────────────────────────────────────────────────╮                                          
┃ │ Function execution error: unhandled errors in a TaskGroup (1 sub-exception)  │                                          
┃ ╰──────────────────────────────────────────────────────────────────────────────╯                                          
  ✔ Container.from(address: "node:21-slim"): Container! 0.6s
  ✔ Container.withExec(args: ["npm", "install"]): Container! 43.2s
  ✘ Container.withExec(args: ["npm", "run", "lint"]): Container! 53.4s
  ┃                                                                                                                         
  ┃ > hello-dagger@0.0.0 lint                                                                                               
  ┃ > eslint . --ext .vue,.js,.jsx,.cjs,.mjs,.ts,.tsx,.cts,.mts --fix --ignore-path .gitignore --max-warnings=0             
  ┃                                                                                                                         
  ┃                                                                                                                         
  ┃ /src/src/App.vue                                                                                                        
  ┃   2:10  warning  'RouterLink' is defined but never used  @typescript-eslint/no-unused-vars                              
  ┃   2:22  warning  'RouterView' is defined but never used  @typescript-eslint/no-unused-vars                              
  ┃                                                                                                                         
  ┃ ✖ 2 problems (0 errors, 2 warnings)                                                                                     
  ┃                                                                                                                         
  ┃ ESLint found too many warnings (maximum: 0).                                                                            
  ✘ Container.stdout: String! 53.6s
  ! process "npm run lint" did not complete successfully: exit code: 1
  ✘ Container.withExec(args: ["npm", "run", "type-check"]): Container! 53.5s
  ┃                                                                                                                         
  ┃ > hello-dagger@0.0.0 type-check                                                                                         
  ┃ > vue-tsc --build --force                                                                                               
  ✘ Container.stdout: String! 53.7s
  ! context canceled
  ✔ Container.withExec(args: ["npm", "run", "test:unit", "run"]): Container! 52.5s
  ✔ Container.stdout: String! 52.7s

Full trace at https://dagger.cloud/dagger/traces/8a500c8bfb5ac0b6d85f38ab1e0cbd8a

Error: response from query: input: myModule.runAllTests resolve: call function "run_all_tests": process "/runtime" did not complete successfully: exit code: 1

Stderr:
╭─ Error ──────────────────────────────────────────────────────────────────────╮
│ Function execution error: unhandled errors in a TaskGroup (1 sub-exception)  │
╰──────────────────────────────────────────────────────────────────────────────╯
```